src/Microsoft.Graph.Core/Authentication/TokenCredentials/IntegratedWindowsTokenCredential.cs

```diff
@@ -0,0 +1,138 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph
+{
+    using Azure.Core;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using System.Collections.Generic;
+    using Microsoft.Identity.Client;
+    using System;
+
+    /// <summary>
+    /// A <see cref="TokenCredential"/> implementation using MSAL.Net to acquire token by integrated windows authentication.
+    /// </summary>
+    public class IntegratedWindowsTokenCredential : TokenCredential
+    {
+
+        private static readonly List<string> WellKnownTenants = new List<string>
+        {   AuthConstants.Tenants.Common,
+            AuthConstants.Tenants.Consumers,
+            AuthConstants.Tenants.Organizations
+        };
+
+        private readonly IPublicClientApplication _clientApplication;
+
+        /// <summary>
+        /// Creates a new IntegratedWindowsTokenCredential which will authenticate users with the specified application.
+        /// </summary>
+        /// <param name="publicClientApplication">A <see cref="IPublicClientApplication"/> to pass to <see cref="IntegratedWindowsTokenCredential"/> for authentication.</param>
+        /// <exception cref="ArgumentException"> When a null <see cref="IPublicClientApplication"/> is passed</exception>
+        public IntegratedWindowsTokenCredential(IPublicClientApplication publicClientApplication)
+        {
+            _clientApplication = publicClientApplication ?? throw new ArgumentException(
+                                     string.Format(ErrorConstants.Messages.NullParameter, nameof(publicClientApplication)),
+                                     nameof(publicClientApplication));
+        }
+
+        /// <summary>
+        /// Gets an <see cref="AccessToken"/> using the provided <see cref="TokenRequestContext"/> in a synchronous fashion
+        /// </summary>
+        /// <param name="requestContext">The <see cref="TokenRequestContext"/> for the request</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> for the request</param>
+        /// <returns>An <see cref="AccessToken"/> to make requests with</returns>
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            return GetTokenAsync(requestContext,cancellationToken).GetAwaiter().GetResult();
```

Comment:
wrapping in a try catch to unwrap the TaskException and return throw the source exception instead might be required here (part of what await does besides waiting for the task to complete)

Reply:
Isn't that the case when you expect an `AggregateException` from calling .Result directly?
https://stackoverflow.com/questions/34793684/catching-original-exception-when-using-async-synchronously

Reply:
interesting, I didn't know this subtilty between .Result and .GetResult()! Thanks for sharing. Resolving.